### PR TITLE
grab-point adjustments

### PR DESCRIPTION
**Pull request by sethalves**
- don't show grab-point spheres if hand-controllers aren't in use
- hand lasers all line up again (including those used on 2D UI)
- handControllerGrab.js no longer backs up a bit before doing a ray-pick against entities
- code that tries to adjust release velocity of a thrown entity has been disabled
- grab-point spheres have been made smaller and their positions have been adjusted
- when a grab-point sphere touches a grabbable entity, the controller will do a haptic pulse
- far-grabbed entities that are held very still before being released will now fall

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 43c98b3f1e634f172bcc281f83774769f8546ddd available at: https://gist.github.com/2c2b75094933e1201537a5769a573e05<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-43c98b3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-43c98b3.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 215fd4ddfdb943ae0456fa619061f59b6d032990 available at: https://gist.github.com/2c2b75094933e1201537a5769a573e05<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-215fd4d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-215fd4d.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 34e4b4ae1964a249b34b3c83580b03caecd1344d available at: https://gist.github.com/2c2b75094933e1201537a5769a573e05<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-34e4b4a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-34e4b4a.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 616b094111435c141cec9469f1e052fb55cc702f available at: https://gist.github.com/2c2b75094933e1201537a5769a573e05<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-616b094.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-616b094.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9ee27ed7d48e3384220f76bd2ecf9cb1b71e5d6c available at: https://gist.github.com/2c2b75094933e1201537a5769a573e05<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-9ee27ed.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-9ee27ed.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 746d0cd91ce2c0b145ee51700d31f0d3081a2713 available at: https://gist.github.com/2c2b75094933e1201537a5769a573e05<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-746d0cd.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-746d0cd.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9c96ffc9cb42ff42274b29f7687f81878ccc8309 available at: https://gist.github.com/2c2b75094933e1201537a5769a573e05<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-9c96ffc.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-9c96ffc.dmg

**Comment by sethalves** [comment]
to test:
- exercise all aspects of grabbing and equipping
- use 2D UI with hand controllers

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9c96ffc9cb42ff42274b29f7687f81878ccc8309 available at: https://gist.github.com/2c2b75094933e1201537a5769a573e05<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-9c96ffc.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-9c96ffc.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9e99cc1f7879156d51af33f759be3170c1303499 available at: https://gist.github.com/2c2b75094933e1201537a5769a573e05<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-9e99cc1.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-9e99cc1.dmg

**Comment by Nex-Pro** [comment]
Tested on win10 with vive.

Grabbing seems comparable to production but the grab-point spheres are about 15 cm off the hand for the mathew avatar.

Initially had huge problems grabbing in dev-demo, basically useless, but after some time spent cleaning the area of scripted junk the grabbing recovered.


don't show grab-point spheres if hand-controllers aren't in use: Success

hand lasers all line up again (including those used on 2D UI): Success

handControllerGrab.js no longer backs up a bit before doing a ray-pick against entities: Success

code that tries to adjust release velocity of a thrown entity has been disabled: Seems better I scored some hoops.


grab-point spheres have been made smaller and their positions have been adjusted: to me this is a fail as the spheres are way off the hand.

when a grab-point sphere touches a grabbable entity, the controller will do a haptic pulse: Success

far-grabbed entities that are held very still before being released will now fall: Success


Still some weirdness when far grabbed object gets close to the grap sphere. Unable to push object away by arm movement.


Release of a grabbed object seems delayed enough to make shooting hoops a miserable experiment in working out how early the release should be made.


Needs testing with hydras.

**Comment by kevinmthomasse** [comment]
testing

**Comment by kevinmthomasse** [comment]
## QA Test
**QA Tester**: KevinMThomas
**QA Pass**: Failure
**Additional Test Results + Observations**:  Upon loading interface and entering into ```dev-demo``` a crash occurred, bug splat sent.  Upon second attempt, frame rate is dramatically low at 11FPS and equipping items is very clumsy and unreliable with the Hydras.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8c45c115b87f811d85d55914b6559bdb2b338032 available at: https://gist.github.com/2c2b75094933e1201537a5769a573e05<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-8c45c11.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-8c45c11.dmg

**Comment by sethalves** [comment]
I've attempted to keep far-grabbed things from getting bogged down after they've been brought close to the grab-point.

The blue grab-point sphere being out in front of the hand is by design -- it's likely we'll provide some sort of setting to adjust this, but we haven't made this yet.

**Comment by Nex-Pro** [comment]
@sethalves "The blue grab-point sphere being out in front of the hand is by design" Doesn't seem right when you are interacting with small objects. Don't want to see the blue ball and want to grab with the hand. I tried shooting hoops and it was miserable to have to grab the ball so it wasn't 'in the hand'.
The bogging down issue is horrible and needs to be gone. 

@kevinmthomasse The low frame rate at dev-demo may have been due to the presence of a lot of scripted junk. When I first started the test there I was getting terrible frame rates. So I cleaned up a lot of wayward objects including those pesky bow-strings and maraccas. My frame rate then recovered to decent levels. I had asked for a clean up after I did the manual sweep and one of the devs ( who shall remain nameless ) restored dev-demo to the mess it was in. lol. So it may be that you were in there at that time.

**Comment by Nex-Pro** [comment]
@sethalves and @ZappoMan I think you're going to need to do some testing where we throw in hundreds of scripted objects from the market and get some metrics on the impact to track down the worst offenders.( You probably already are )

File: interface/src/avatar/AvatarActionHold.cpp [hunk 1]
```diff
@@ -244,14 +244,18 @@ void AvatarActionHold::doKinematicUpdate(float deltaTimeStep) {
...
-                (i + 2) % AvatarActionHold::velocitySmoothFrames == _measuredLinearVelocitiesIndex ||
-                (i + 3) % AvatarActionHold::velocitySmoothFrames == _measuredLinearVelocitiesIndex) {
-                continue;
-            }
+
+            // if ((i + 1) % AvatarActionHold::velocitySmoothFrames == _measuredLinearVelocitiesIndex ||
+            //     (i + 2) % AvatarActionHold::velocitySmoothFrames == _measuredLinearVelocitiesIndex ||
+            //     (i + 3) % AvatarActionHold::velocitySmoothFrames == _measuredLinearVelocitiesIndex) {
+            //     continue;
+            // }
```
Comment (AndrewMeadows) [inline interface/src/avatar/AvatarActionHold.cpp]: Is this cruft is being kept around because we might resurrect it?

Comment (sethalves) [inline interface/src/avatar/AvatarActionHold.cpp]: Yes, I suspect it will be back.  I'll add a comment saying so.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 59db640cb54c723d0c2011534663281978311d9b available at: https://gist.github.com/2c2b75094933e1201537a5769a573e05<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-59db640.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8571-59db640.dmg

**Pull request merged**